File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.restapi.PlaceOrderRequest;
import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.drivers.TimeServiceDriver;
import com.wimp.app.specs.support.OrderData;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

import java.time.LocalTime;
import java.util.List;

public class OrderingStepDefinitions {
    private final OrderingApiDriver orderingApiDriver;
    private final TimeServiceDriver timeServiceDriver;

    public OrderingStepDefinitions(OrderingApiDriver orderingApiDriver, TimeServiceDriver timeServiceDriver) {
        this.orderingApiDriver = orderingApiDriver;
        this.timeServiceDriver = timeServiceDriver;
    }

    @Given("they have placed an order with")
    public void theyHavePlacedAnOrderWith(List<OrderData> orderDataList) throws Exception {
        var orderData = orderDataList.getFirst(); // a single-row data table is expected for this step

        // ensuring that the placing time is before the expected delivery time
        timeServiceDriver.setCurrentTime(orderData.getExpectedDeliveryTime().minusMinutes(5));

        // preparing a place order request with expected delivery time (this setting is only available for testing)
        PlaceOrderRequest placeOrderRequest = new PlaceOrderRequestObjectMother()
            .withExpectedDeliveryTime(timeServiceDriver.getTodayTime(orderData.getExpectedDeliveryTime()))
            .build();

        orderingApiDriver.placeOrder(placeOrderRequest).execute();
    }

    @When("the delivery has not been made by {time}")
    public void whenTheDeliveryHasNotBeenMadeBy(LocalTime time) {
        timeServiceDriver.setCurrentTime(time);
    }
}

File: Reqnroll/Chapter15/15.7-StubDependencyPattern/After/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(TimeServiceDriver timeServiceDriver, OrderingApiDriver orderingApiDriver)
{
    public record OrderRequestData(TimeSpan ExpectedDeliveryTime);

    [Given("they have placed an order with")]
    public async Task GivenTheyHavePlacedAnOrderWith(DataTable orderDataTable)
    {
        var orderData = orderDataTable.CreateInstance<OrderRequestData>();
        var expectedDeliveryTime = TimeOnly.FromTimeSpan(orderData.ExpectedDeliveryTime);

        // ensuring that the placing time is before the expected delivery time
        timeServiceDriver.SetCurrentTime(expectedDeliveryTime.Add(TimeSpan.FromMinutes(-5)));
        // preparing a place order request with expected delivery time (this setting is only available for testing)
        var placeOrderRequest = new PlaceOrderRequestObjectMother()
            .WithExpectedDeliveryTime(timeServiceDriver.GetTodayTime(expectedDeliveryTime))
            .Build();

        await orderingApiDriver.PlaceOrder(placeOrderRequest).Execute();
    }

    [When("the delivery has not been made by {TimeOnly}")]
    public void WhenTheDeliveryHasNotBeenMadeBy(TimeOnly time)
    {
        timeServiceDriver.SetCurrentTime(time);
    }
}
